# Prototype: Object Detection for Autonomous Vehicles in Inclement Weather

<img src="images/test_5.jpg" alt="Description or alt text for the image" title="Optional title" width="600"/>

Figure 1: Example of Object Detection with Inclement Weather

## 1. Introduction

### 1.1 Dataset and Augmentation Strategy
To advance our object detection prototype, we've curated a specialized dataset, subsequently enriched through strategic augmentation to bolster model training and performance. Leveraging **Roboflow** for its robust image augmentation capabilities, we enhanced our dataset, initially comprising 198 images, by introducing realistic variations like blur (up to 1.7px) and noise (affecting up to 0.3% of pixels). This approach aimed to simulate the challenges inherent in inclement weather conditions, directly impacting autonomous vehicle navigation.

**Augmented Dataset Composition:**
- **Training Set**: Enriched to 278 images for comprehensive learning.
- **Validation Set**: 39 images for parameter optimization.
- **Test Set**: 20 images for evaluating model efficacy.

**Label Distribution:**
- **Cars**: Predominantly featured with 1,178 labels.
- **Trucks**: Represented by 134 labels.
- **Persons**: Included with 278 labels to mimic real-world scenarios.

### 1.2 Fine-tuning Strategy

#### Batch Size
- **Evaluation**: Tested with batch sizes 8 and 16, acknowledging the default of 16 as optimal.
- **Justification**: Balances computational demands with learning efficiency, crucial for training on augmented datasets without compromising on model training effectiveness.

#### Learning Rate Scheduler
- **Adoption**: Utilized a cosine learning rate scheduler, enhancing the default setting.
- **Advantage**: Facilitates better model performance through adaptive learning rate adjustments, avoiding local minima and refining model weights more precisely over time.

#### Dataset Expansion
- **Implementation**: Nearly doubled the dataset size, integrating data augmentation to introduce a wider array of visual scenarios.
- **Weather Conditions**: Dataset inlcudes cloudy and snowy weather conditions.
- **Objective**: To enhance the model's robustness and its ability to generalize, ensuring reliable performance under diverse environmental conditions, including inclement weather.

## 2. Experiment Overview and Configurations for Inclement Weather Object Detection

To advance the capabilities of our object detection model specifically designed for inclement weather conditions, we conducted a series of experiments focusing on the YOLOv8 trained model. The objective was to fine-tune the model's settings—particularly the learning rate scheduler and batch size—to identify the optimal configuration that ensures both high accuracy and computational efficiency.

### 2.1. Experiment Configurations

### Learning Rate Scheduler
- **Description**: Adjusts whether the cosine learning rate scheduler is active, dynamically changing the learning rate in a cosine pattern throughout training.
- **Options**: Enabled (`True`) or Disabled (`False`).
- **Goal**: Investigate the scheduler's role in enhancing the model's ability to generalize and avoid local minima, crucial for robust detection in varying weather conditions.

### Batch Size
- **Description**: Specifies the number of training examples the model processes in a single iteration.
- **Options**: `8` and `16`.
- **Goal**: Assess the impact of batch size on the learning process and computational load, aiming to find a balance that optimizes both learning efficiency and resource usage.

### Conducted Experiments

Four experiments were designed with different configurations to explore how each affects the performance of the model in detecting objects under inclement weather:

1. **Experiment 1 - 8T**: Batch Size of 8, Learning Rate Scheduler Enabled.
2. **Experiment 2 - 8F**: Batch Size of 8, Learning Rate Scheduler Disabled.
3. **Experiment 3 - 16T**: Batch Size of 16, Learning Rate Scheduler Enabled.
4. **Experiment 4 - 16F**: Batch Size of 16, Learning Rate Scheduler Disabled.

The aim was to isolate the effects of these two parameters on the model's accuracy and processing efficiency, providing insights into the most effective training regimen for inclement weather object detection.

### 2.2. Objective

The primary goal of these experiments is to fine-tune the YOLOv8 model for enhanced performance in inclement weather scenarios. By optimizing training parameters, we strive to improve the model's precision in detecting objects amidst challenging weather conditions, an essential capability for various applications, including autonomous vehicle navigation and outdoor surveillance.

### Results Summary

The detailed training and validation metrics for each experiment will be discussed later in this notebook. This section is dedicated to summarizing and interpreting the outcomes, focusing on their implications for developing a more accurate and efficient object detection system for inclement weather. The comprehensive analysis aims to facilitate a straightforward understanding of the experimental findings and their relevance to the project's objectives.


## 3. Results and Analysis

The fine-tuning of our YOLOv8 Small model, with a focus on batch size and learning rate adjustments, led to significant improvements in performance metrics. This section delves into the detailed outcomes of our experiments, particularly highlighting the performance of the model configuration labeled as "16T" - which denotes a batch size of 16 and an enabled cosine learning rate scheduler.

<img src="images/confusion_matrix_normalized.png" alt="Description or alt text for the image" title="Optional title" width="700"/>
Figure 2: Confusion Matrix Normalized.

### 3.1.Analysis of Normalized Confusion Matrix Results

The normalized confusion matrix provides a detailed view of the model's classification performance, with a focus on the proportion of correct and incorrect predictions for each class, normalized by the number of true instances for those classes. This perspective is particularly valuable for understanding the model's behavior in a balanced manner, especially when dealing with imbalanced datasets.

### Key Observations from the Normalized Confusion Matrix

**Analysis of the Confusion Matrix Results**

The confusion matrix offers invaluable insights into the performance of our object detection model, particularly regarding its ability to accurately classify different objects within flood scenarios. Here's a detailed analysis based on the diagonal and off-diagonal values observed in the matrix:

**High Diagonal Values**
- The diagonal values of the confusion matrix represent the model's prediction accuracy for each class. Notably, the **truck** class achieves the highest accuracy with a value of **0.86**, indicating exceptional model performance in recognizing trucks. This is followed by the **person** class with an accuracy of **0.83**, and the **car** class with an accuracy of **0.8**. These findings suggest a hierarchy in detection capabilities, with trucks being identified most reliably.

**Off-diagonal Values and Misclassification Trends**
- Off-diagonal values reveal instances where the model confuses one class for another. A significant observation is that the **car** class is frequently misclassified as **background**, leading to missed detections. This misclassification significantly affects the car's lower accuracy despite being the predominant label in our dataset. Furthermore, the **person** class is often misclassified as **background** as well.

<img src="images/results.png" alt="Description or alt text for the image" title="Optional title" width="700"/>

Figure 3: mAP and loss plots after training the YOLOv8 Small model on the dataset.

<img src="images/train_batch2.jpg" alt="Description or alt text for the image" title="Optional title" width="700"/>

Figure 4: Example training batch 2 of the dataset.

### 3.2. Detailed Performance Metrics for 16T Configuration

The 16T configuration yielded the following enhanced performance metrics, showcasing notable improvements across various aspects:

- **Precision (Box(P))**: Achieved a high precision rate of 86.2%, indicating the model's accuracy in predicting relevant objects across all classes. This metric is particularly crucial for autonomous vehicles, where false positives can lead to unnecessary or hazardous maneuvers.
  
- **Recall (R)**: Recorded at 77.5%, this metric reflects the model's ability to correctly identify all relevant instances in the dataset. Although there's a slight decrease from the pre-fine-tuning state, the trade-off is balanced by substantial gains in precision.

- **mAP0.5**: With an mAP of 84.3%, the model demonstrates strong consistency in detecting objects accurately at the 0.5 IoU threshold. This improvement is indicative of the model's robustness in various detection scenarios.

- **mAP0.5:0.95**: The increase to 58.9% in this metric highlights the model's improved performance across a range of IoU thresholds, showcasing its versatility and reliability in more precise object detection tasks.

### Class-Specific Performance Insights

- **Cars**: The model maintained high performance in car detection, a critical aspect for navigating traffic and urban environments, with a precision of 85.8% and an mAP0.5 of 86.1%.
  
- **Trucks**: Notably, truck detection saw a remarkable improvement, with precision jumping to 94.7% and mAP0.5 reaching 86.9%. This enhancement underscores the model's increased capability in identifying larger vehicles, vital for ensuring safety on highways and in mixed-traffic situations.
  
- **Persons**: The introduction of person detection with a precision of 78.1% and an mAP0.5 of 80.0% marks a significant step forward. Despite being a newly added class, the model's ability to detect pedestrians enhances its application in ensuring pedestrian safety around autonomous vehicles.


### Key Highlights
- **Precision** improved overall from 75.9% to 86.2%, indicating higher accuracy in the model's predictions.
- **Recall** experienced a minor reduction from 79.2% to 77.5%, a slight trade-off for the gain in precision.
- **Mean Average Precision (mAP50)** increased from 80.6% to 84.3%, with **mAP50-95** also rising from 54.2% to 58.9%, demonstrating enhanced detection reliability across varying Intersection over Union (IoU) thresholds.


### 3.3. Reflection on Dataset Imbalance and Model Performance

The imbalanced nature of the dataset, with a predominant focus on cars, necessitates careful consideration. While the model demonstrates great ability in detecting vehicles, the relatively lower representation of trucks and persons could influence its learning bias:

- **Oversampling Underrepresented Classes**: By artificially increasing the number of instances of underrepresented classes in the training dataset, we can ensure that the model receives adequate exposure to these classes, reducing the risk of overlooking them.

- **Advanced Data Augmentation Techniques**: Employing a variety of data augmentation techniques such as random rotations, scaling, cropping, and color adjustments can introduce more variability into the training data. This not only helps in addressing class imbalance but also in improving the model's robustness to diverse conditions.

- **Adjusting Class Weights**: During the training phase, assigning higher weights to underrepresented classes can compel the model to pay more attention to minimizing errors on these classes. This adjustment helps in balancing the influence of each class on the model's learning process.

- **Expanding the Dataset with More Diverse Scenarios**: In addition to the above strategies, expanding the dataset to include a broader range of scenarios for all three classes—cars, trucks, and persons—can introduce more diversity into the model's training regime. Incorporating images from different weather conditions, times of day ensures that the model learns to recognize these classes across a wide spectrum of real-world situations. This expansion not only addresses class imbalance but also enhances the model's generalization capabilities, making it more effective in accurately detecting objects in complex driving scenarios.

By combining these strategies, we can significantly improve the model's ability to detect and classify objects with greater accuracy and fairness across all classes. This holistic approach to training will pave the way for developing a highly reliable object detection system for autonomous vehicles.

## Conclusion
The fine-tuned YOLOv8 Small model, supported by an augmented and expanded dataset, showcases promising advancements in object detection crucial for autonomous vehicles in inclement weather. This project exemplifies the importance of targeted dataset preparation, strategic parameter adjustment, and the ongoing need to address data imbalances for optimal model performance.


## 4. Experiment Details

### 4.1. Experiment Configurations

- **Learning Rate Scheduler**: Indicates whether the cosine learning rate scheduler was enabled (`True`) or disabled (`False`). The scheduler adjusts the learning rate in a cosine pattern, potentially aiding in avoiding local minima and achieving better generalization.
- **Batch Size**: The number of training examples utilized in one iteration. We explored sizes of `8` and `16` to understand their impact on the model's learning dynamics and computational efficiency.


In [18]:
# !pip install roboflow

from roboflow import Roboflow
rf = Roboflow(api_key="IvxuC5AtkVBPx3o8pcGV")
project = rf.workspace("machine-learning-engineer-bootcamp").project("object_detect_proto")
dataset = project.version(2).download("yolov8")


loading Roboflow workspace...
loading Roboflow project...
Dependency ultralytics==8.0.196 is required but found version=8.0.225, to fix: `pip install ultralytics==8.0.196`



Extracting Dataset Version Zip to object_detect_proto-2 in yolov8:: 100%|██████████| 686/686 [00:00<00:00, 826.73it/s]


In [36]:
# !pip install ultralytics
from ultralytics import YOLO
import ultralytics
ultralytics.checks()

Ultralytics YOLOv8.0.225 🚀 Python-3.11.4 torch-2.1.1+cpu CPU (Intel Core(TM) i7-7600U 2.80GHz)
Setup complete ✅ (4 CPUs, 15.8 GB RAM, 239.6/953.3 GB disk)


### Experiment 1 - 8T (Batch Size: 8, Learning Rate Scheduler: True)

In [21]:
%%capture captured_output
# Code that produces output

model_1 = YOLO('yolov8s.pt')  # load a pretrained model 
# Train the model
results = model_1.train(data='data.yaml', epochs=50, imgsz=640, batch= 8, cos_lr=True)

New https://pypi.org/project/ultralytics/8.1.11 available 😃 Update with 'pip install -U ultralytics'
Ultralytics YOLOv8.0.225 🚀 Python-3.11.4 torch-2.1.1+cpu CPU (Intel Core(TM) i7-7600U 2.80GHz)
engine\trainer: task=detect, mode=train, model=yolov8s.pt, data=data.yaml, epochs=50, patience=50, batch=8, imgsz=640, save=True, save_period=-1, cache=False, device=None, workers=8, project=None, name=train8, exist_ok=False, pretrained=True, optimizer=auto, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=True, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=None, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, vid_stride=1, stream_buffer=False, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, show=False, save_frames=False, save_txt=False, save_conf=False, save_cro


      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
                   all         39        290      0.822      0.609      0.718      0.424

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
                   all         39        290      0.812      0.617      0.717      0.419

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
                   all         39        290      0.837      0.585      0.722      0.424

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
                   all         39        290      0.847      0.609      0.742      0.455

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
                   all         39        290      0.856      0.591      0.753      0.476

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
                   all         39        290       0.78      0.679       0.76    

In [22]:
metrics = model_1.val()
metrics.box.map    # map50-95
metrics.box.map50  # map50
metrics.box.map75  # map75
metrics.box.maps   # a list contains map50-95 of each category

Ultralytics YOLOv8.0.225 🚀 Python-3.11.4 torch-2.1.1+cpu CPU (Intel Core(TM) i7-7600U 2.80GHz)
Model summary (fused): 168 layers, 11126745 parameters, 0 gradients, 28.4 GFLOPs


val: Scanning C:\Users\rafae\Documents\Capstone\object_detect_proto-2\valid\labels.cache... 39 images, 6 backgrounds, 0 corrupt: 100%|██████████| 39/39 [00:00<?, ?it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 5/5 [00:31<00:00,  6.35s/it]


                   all         39        290      0.872      0.695      0.815      0.559
                   car         39        221      0.906      0.692      0.851      0.598
                person         39         48      0.804      0.583      0.713      0.382
                 truck         39         21      0.905       0.81      0.882      0.697
Speed: 6.6ms preprocess, 779.9ms inference, 0.0ms loss, 1.1ms postprocess per image
Results saved to runs\detect\train82


array([    0.59774,     0.38205,      0.6966])

### Experiment 2 - 8F (Batch Size: 8, Learning Rate Scheduler: False)

In [23]:
%%capture captured_output
# Code that produces output

model_2 = YOLO('yolov8s.pt')  # load a pretrained model 
# Train the model
results = model_2.train(data='data.yaml', epochs=50, imgsz=640, batch= 8, cos_lr=False)

New https://pypi.org/project/ultralytics/8.1.11 available 😃 Update with 'pip install -U ultralytics'
Ultralytics YOLOv8.0.225 🚀 Python-3.11.4 torch-2.1.1+cpu CPU (Intel Core(TM) i7-7600U 2.80GHz)
engine\trainer: task=detect, mode=train, model=yolov8s.pt, data=data.yaml, epochs=50, patience=50, batch=8, imgsz=640, save=True, save_period=-1, cache=False, device=None, workers=8, project=None, name=train9, exist_ok=False, pretrained=True, optimizer=auto, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=None, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, vid_stride=1, stream_buffer=False, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, show=False, save_frames=False, save_txt=False, save_conf=False, save_cr


      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
                   all         39        290      0.772      0.597      0.718       0.43

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
                   all         39        290      0.821      0.609      0.732      0.427

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
                   all         39        290      0.857      0.615      0.754      0.451

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
                   all         39        290      0.762       0.68      0.748      0.475

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
                   all         39        290      0.816      0.654      0.778      0.485

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
                   all         39        290      0.751      0.643      0.752    

In [24]:
metrics = model_2.val()
metrics.box.map    # map50-95
metrics.box.map50  # map50
metrics.box.map75  # map75
metrics.box.maps   # a list contains map50-95 of each category

Ultralytics YOLOv8.0.225 🚀 Python-3.11.4 torch-2.1.1+cpu CPU (Intel Core(TM) i7-7600U 2.80GHz)
Model summary (fused): 168 layers, 11126745 parameters, 0 gradients, 28.4 GFLOPs


val: Scanning C:\Users\rafae\Documents\Capstone\object_detect_proto-2\valid\labels.cache... 39 images, 6 backgrounds, 0 corrupt: 100%|██████████| 39/39 [00:00<?, ?it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 5/5 [00:38<00:00,  7.61s/it]


                   all         39        290      0.824      0.705       0.81      0.565
                   car         39        221      0.881      0.702      0.861      0.616
                person         39         48      0.751      0.604      0.708      0.398
                 truck         39         21      0.841       0.81      0.862       0.68
Speed: 7.6ms preprocess, 936.2ms inference, 0.0ms loss, 1.2ms postprocess per image
Results saved to runs\detect\train92


array([     0.6163,     0.39814,      0.6801])

### Experiment 3 - 16T (Batch Size: 16, Learning Rate Scheduler: True)

In [25]:
%%capture captured_output
# Code that produces output

model_3 = YOLO('yolov8s.pt')  # load a pretrained model 
# Train the model
results = model_3.train(data='data.yaml', epochs=50, imgsz=640, batch= 16, cos_lr=True)

New https://pypi.org/project/ultralytics/8.1.11 available 😃 Update with 'pip install -U ultralytics'
Ultralytics YOLOv8.0.225 🚀 Python-3.11.4 torch-2.1.1+cpu CPU (Intel Core(TM) i7-7600U 2.80GHz)
engine\trainer: task=detect, mode=train, model=yolov8s.pt, data=data.yaml, epochs=50, patience=50, batch=16, imgsz=640, save=True, save_period=-1, cache=False, device=None, workers=8, project=None, name=train10, exist_ok=False, pretrained=True, optimizer=auto, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=True, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=None, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, vid_stride=1, stream_buffer=False, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, show=False, save_frames=False, save_txt=False, save_conf=False, save_c


      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
                   all         39        290      0.836      0.634      0.746      0.431

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
                   all         39        290      0.839      0.598      0.746      0.469

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
                   all         39        290      0.816      0.633       0.75      0.459

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
                   all         39        290      0.787      0.672      0.756       0.47

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
                   all         39        290      0.842      0.648      0.779      0.475

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
                   all         39        290      0.846      0.672      0.791    

In [26]:
metrics = model_3.val()
metrics.box.map    # map50-95
metrics.box.map50  # map50
metrics.box.map75  # map75
metrics.box.maps   # a list contains map50-95 of each category

Ultralytics YOLOv8.0.225 🚀 Python-3.11.4 torch-2.1.1+cpu CPU (Intel Core(TM) i7-7600U 2.80GHz)
Model summary (fused): 168 layers, 11126745 parameters, 0 gradients, 28.4 GFLOPs


val: Scanning C:\Users\rafae\Documents\Capstone\object_detect_proto-2\valid\labels.cache... 39 images, 6 backgrounds, 0 corrupt: 100%|██████████| 39/39 [00:00<?, ?it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:32<00:00, 10.71s/it]


                   all         39        290      0.862      0.775      0.843      0.589
                   car         39        221      0.858      0.765      0.861      0.608
                person         39         48      0.781      0.708        0.8      0.457
                 truck         39         21      0.947      0.852      0.869      0.702
Speed: 6.5ms preprocess, 792.8ms inference, 0.0ms loss, 1.0ms postprocess per image
Results saved to runs\detect\train102


array([    0.60849,     0.45694,      0.7024])

### Experiment 4 - 16F (Batch Size: 16, Learning Rate Scheduler: False)

In [27]:
%%capture captured_output
# Code that produces output

model_4 = YOLO('yolov8s.pt')  # load a pretrained model 
# Train the model
results = model_4.train(data='data.yaml', epochs=50, imgsz=640, batch= 16, cos_lr=False)

New https://pypi.org/project/ultralytics/8.1.11 available 😃 Update with 'pip install -U ultralytics'
Ultralytics YOLOv8.0.225 🚀 Python-3.11.4 torch-2.1.1+cpu CPU (Intel Core(TM) i7-7600U 2.80GHz)
engine\trainer: task=detect, mode=train, model=yolov8s.pt, data=data.yaml, epochs=50, patience=50, batch=16, imgsz=640, save=True, save_period=-1, cache=False, device=None, workers=8, project=None, name=train11, exist_ok=False, pretrained=True, optimizer=auto, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=None, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, vid_stride=1, stream_buffer=False, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, show=False, save_frames=False, save_txt=False, save_conf=False, save_


      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
                   all         39        290      0.813      0.662      0.744      0.443

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
                   all         39        290      0.736       0.69      0.745       0.43

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
                   all         39        290      0.767      0.673      0.768      0.464

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
                   all         39        290      0.884      0.694      0.782      0.489

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
                   all         39        290       0.84      0.699      0.804      0.518

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
                   all         39        290       0.88      0.664      0.772    

In [28]:
metrics = model_4.val()
metrics.box.map    # map50-95
metrics.box.map50  # map50
metrics.box.map75  # map75
metrics.box.maps   # a list contains map50-95 of each category

Ultralytics YOLOv8.0.225 🚀 Python-3.11.4 torch-2.1.1+cpu CPU (Intel Core(TM) i7-7600U 2.80GHz)
Model summary (fused): 168 layers, 11126745 parameters, 0 gradients, 28.4 GFLOPs


val: Scanning C:\Users\rafae\Documents\Capstone\object_detect_proto-2\valid\labels.cache... 39 images, 6 backgrounds, 0 corrupt: 100%|██████████| 39/39 [00:00<?, ?it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:33<00:00, 11.08s/it]


                   all         39        290      0.822      0.729      0.822      0.586
                   car         39        221      0.822      0.733      0.849      0.628
                person         39         48      0.802      0.646      0.752      0.428
                 truck         39         21      0.843       0.81      0.864      0.701
Speed: 6.8ms preprocess, 823.1ms inference, 0.0ms loss, 0.8ms postprocess per image
Results saved to runs\detect\train112


array([    0.62796,     0.42754,     0.70107])

### 4.2. Comparative Analysis

The experiments conducted to optimize our YOLOv8 model for inclement weather object detection provided valuable insights into the impact of learning rate scheduler activation and batch size variations. Below, we summarize the findings from each experiment and explain our decision to select the 16T configuration as the optimal setup.

### Summary of Results

- **Experiment 1 - 8T**:
  - Duration: 9.235 hours
  - Highest accuracy for truck detection (.882 mAP50), but overall performance for all classes (.815 mAP50) not optimal.
  - Slightly faster inference time compared to other configurations.

- **Experiment 2 - 8F**:
  - Duration: 9.403 hours
  - Moderate performance across all classes with a slight improvement in mAP50-95 scores compared to 8T.
  - Longer inference time, indicating less efficiency.

- **Experiment 3 - 16T** (Selected Configuration):
  - Duration: 9.904 hours
  - Showed the best balance between accuracy and efficiency, with the highest overall performance (.843 mAP50) and significant improvement in truck detection (.869 mAP50).
  - Slightly longer training duration but compensated with improved performance and efficient inference time.

- **Experiment 4 - 16F**:
  - Duration: 9.794 hours
  - Comparable training duration to 16T but with slightly reduced performance metrics across all classes.
  - Inference speed and postprocess time did not justify the slight decrease in performance metrics.

### Explanation for Selecting 16T Configuration 

The decision to proceed with the **16T configuration** as the prototype model for our project was based on a comprehensive evaluation of the trade-offs between training duration, accuracy across classes, and inference efficiency. Here's why 16T stood out:

- **Optimal Performance**: The 16T configuration demonstrated superior ability to accurately detect objects across all tested classes, especially for trucks which are critical for inclement weather conditions. The improvement in precision and mAP scores for truck detection underlines the model's enhanced capability to differentiate between objects crucial for navigating adverse weather.

- **Balanced Efficiency**: While the 16T experiment took slightly longer to complete training, the negligible increase in duration is justified by notable improvements in detection accuracy and model responsiveness. The balance between training time and performance metrics suggests a more robust model capable of delivering reliable results in real-world applications.

- **Generalization and Inference Speed**: The enabled cosine learning rate scheduler in the 16T configuration likely contributed to better generalization, reducing the risk of overfitting and enhancing the model's adaptability to diverse inclement weather scenarios. Moreover, the efficient inference time ensures that the model can process images swiftly, an essential feature for real-time object detection applications.

## Conclusion

Considering the critical importance of accuracy and efficiency in object detection for inclement weather, the 16T configuration emerges as the most suitable choice for our prototype. It strikes an optimal balance between high detection accuracy, particularly for critical classes like trucks, and practical operational efficiency. Future work will focus on further refining this model to enhance its reliability and performance in detecting objects under various inclement weather conditions.


## References

- **YOLOv8 Predict Documentation**: [Ultralytics YOLOv8 Predict Documentation](https://docs.ultralytics.com/modes/predict/#inference-arguments) provides detailed insights on the inference arguments and usage of the YOLOv8 model for predictions.

- **YOLO Resources**:
  - For additional resources and configurations related to YOLO models, visit the Ultralytics GitHub repository: [Ultralytics Default Configuration](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/cfg/default.yaml).

